src/tasks/tasks.hpp

```diff
@@ -0,0 +1,488 @@
+//========================================================================================
+// (C) (or copyright) 2023. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+
+#ifndef TASKS_TASKS_HPP_
+#define TASKS_TASKS_HPP_
+
+#include <algorithm>
+#include <array>
+#include <cassert>
+#include <functional>
+#include <list>
+#include <memory>
+#include <unordered_map>
+#include <unordered_set>
+#include <utility>
+#include <vector>
+
+#include <parthenon_mpi.hpp>
+
+#include "thread_pool.hpp"
+
+namespace parthenon {
+
+enum class TaskListStatus { complete }; // doesn't feel like we need this...
+enum class TaskType { normal, completion };
+
+class TaskQualifier {
+ public:
+  using qualifier_t = uint64_t;
+  TaskQualifier() = delete;
+  TaskQualifier(const qualifier_t n) : flags(n) {} // NOLINT(runtime/explicit)
+
+  static inline constexpr qualifier_t normal{0};
+  static inline constexpr qualifier_t local_sync{1 << 0};
+  static inline constexpr qualifier_t global_sync{1 << 1};
+  static inline constexpr qualifier_t completion{1 << 2};
+  static inline constexpr qualifier_t once_per_region{1 << 3};
+
+  bool LocalSync() const { return flags & local_sync; }
+  bool GlobalSync() const { return flags & global_sync; }
+  bool Completion() const { return flags & completion; }
+  bool Once() const { return flags & once_per_region; }
+  bool Valid() const {
+    if (LocalSync() && GlobalSync()) return false;
+    return true;
+  }
+
+ private:
+  qualifier_t flags;
+};
+
+// forward declare Task for TaskID
+class Task;
+class TaskID {
+ public:
+  TaskID() : task(nullptr) {}
+  // pointers to Task are implicitly convertible to TaskID
+  TaskID(Task *t) : task(t) {} // NOLINT(runtime/explicit)
+
+  TaskID operator|(const TaskID &other) const {
+    // calling this operator means you're building a TaskID to hold a dependency
+    TaskID result;
+    if (task) {
+      result.dep.push_back(task);
+    } else {
+      result.dep.insert(result.dep.end(), dep.begin(), dep.end());
+    }
+    if (other.task) {
+      result.dep.push_back(other.task);
+    } else {
+      result.dep.insert(result.dep.end(), other.dep.begin(), other.dep.end());
+    }
```

Comment:
Why is this `if/else` branching required here? Why not just always add a task if present and allways append dependencies if present? e.g.,
```suggestion
    if (task != nullptr) result.dep.push_back(task);
    result.dep.insert(result.dep.end(), dep.begin(), dep.end());
    if (other.task != nullptr) result.dep.push_back(other.task);
    result.dep.insert(result.dep.end(), other.dep.begin(), other.dep.end());
```
Is there any danger that you could end up in a state where that wouldn't be correct?

Reply:
No I think that should work, and it's cleaner.